lua/vis-std.lua

```diff
@@ -37,34 +37,103 @@ vis:option_register("horizon", "number", function(horizon)
 	return true
 end, "Number of bytes to consider for syntax highlighting")
 
-vis.events.subscribe(vis.events.WIN_HIGHLIGHT, function(win)
-	if not win.syntax or not vis.lexers.load then return end
-	local lexer = vis.lexers.load(win.syntax, nil, true)
-	if not lexer then return end
-
-	-- TODO: improve heuristic for initial style
-	local viewport = win.viewport
-	if not viewport then return end
-	local horizon_max = win.horizon or 32768
-	local horizon = viewport.start < horizon_max and viewport.start or horizon_max
-	local view_start = viewport.start
-	local lex_start = viewport.start - horizon
-	local token_start = lex_start
-	viewport.start = token_start
-	local data = win.file:content(viewport)
-	local token_styles = lexer._TOKENSTYLES
-	local tokens = lexer:lex(data, 1)
-
-	for i = 1, #tokens, 2 do
-		local token_end = lex_start + tokens[i+1] - 1
-		if token_end >= view_start then
+vis:option_register("redrawtime", "string", function(redrawtime)
+	if not vis.win then return false end
+	local value = tonumber(redrawtime)
+	if not value or value <= 0 then
+		vis:info("A positive real number expected")
+		return false
+	end
+	vis.win.redrawtime = value
+	return true
+end, "Number of seconds to wait for syntax highlighting before switching to using the horizon option")
+
+local function populate(horizons, view_start, tokens, horizon_min)
+	for i = 2, #tokens, 2 do
+		if tokens[i] >= view_start then break end
+		if tokens[i] - horizons[#horizons] > horizon_min then
+			table.insert(horizons, tokens[i] - 1)
+		end
+	end
+end
+
+local function purge(list, base)
+	for i = #list, 1, -1 do
+		if list[i] > base then
+			table.remove(list)
+		else
+			break
+		end
+	end
+end
+
+local function make_highlighter()
+	local horizons = {0}
+	local last_view_start = 0
+	local last_view_len = 0
+	local redrawtime_exceeded
+	return function(win)
+		if not win.syntax or not vis.lexers.load then return end
+		local lexer = vis.lexers.load(win.syntax, nil, true)
+		if not lexer then return end
+		local viewport = win.viewport
+		if not viewport then return end
+		local view_start = viewport.start
+		local horizon_min = last_view_len  -- keeps horizon candidates one screenful apart. could be some constant as well
+		local leap = view_start - last_view_start
+		if leap <= last_view_len then
+			purge(horizons, view_start)
+		end
+		local horizon_max = win.horizon or 32768
+		local approx_horizon = view_start < horizon_max and view_start or view_start - horizon_max
+		local lex_start = redrawtime_exceeded and approx_horizon or horizons[#horizons]
+		local data = win.file:content({start = lex_start, finish = viewport.finish})
+		local redrawtime_max = win.redrawtime or 2
+		local tokens, timedout = lexer:lex(data, 1, redrawtime_max)
+		if not redrawtime_exceeded and timedout then
+			redrawtime_exceeded = timedout
+			tokens = lexer:lex(data, 1, redrawtime_max, approx_horizon)
```

Comment:
As mentioned, I don't think trying again makes sense?

Also `approx_horizon` passed as `init` to `lpeg.match` needs to be an index relative to the subject i.e. `data` while here you are calculating with absolute byte offsets?

Reply:
If we don't try again, the user lands on a black-and-white screen when a timeout occurs.
They'll have to move the cursor or something like that to see colors again.
As I mentioned in the other comments, this thing happens only once in the "lifetime" of the open file, so it's not a performance concern.

About the `approx_horizon` - very good catch!
I'll subtract `lex_start`.